tests/numerics/project_04.cc

```diff
@@ -0,0 +1,173 @@
+
+
+// check VectorTools::project_distributed for quadrature data
+
+
+#include "../tests.h"
+#include <deal.II/base/function.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/lac/constraint_matrix.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_refinement.h>
+
+#include <deal.II/lac/generic_linear_algebra.h>
+#include <deal.II/base/quadrature_point_data.h>
+
+#include <fstream>
+
+namespace LA
+{
+  using namespace ::LinearAlgebraTrilinos;
+}
+
+struct QData
+{
+  double density;
+};
+
+
+// define the multi-linear function x or x*y or x*y*z that we will
+// subsequently project onto the ansatz space
+template <int dim>
+class F : public Function<dim>
+{
+public:
+  virtual double value (const Point<dim> &p,
+                        const unsigned int = 0) const
+  {
+    double s = 1;
+    for (unsigned int i=0; i<dim; ++i)
+      s *= p[i];
+    return s;
+  }
+};
+
+const unsigned int fe_degree = 1;
+
+template<int dim>
+void test()
+{
+  F<dim> function;
+
+  parallel::distributed::Triangulation<dim> tria(MPI_COMM_WORLD);
+
+  GridGenerator::hyper_cube (tria);
+  tria.refine_global (2);
+
+  FE_Q<dim> fe(fe_degree);
+  DoFHandler<dim> dof_handler (tria);
+  dof_handler.distribute_dofs (fe);
+
+  QGauss<dim> quadrature_formula(3);
+
+  CellDataStorage<typename parallel::distributed::Triangulation<dim,dim>::cell_iterator,QData> qp_manager;
+  {
+    FEValues<dim> fe_values (fe, quadrature_formula,
+                             update_quadrature_points);
+
+    const unsigned int   n_q_points    = quadrature_formula.size();
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit (cell);
+          qp_manager.initialize(cell, quadrature_formula.size());
+          std::vector<std::shared_ptr<QData> > qpd = qp_manager.get_data(cell);
+          for (unsigned int q=0; q<n_q_points; ++q)
+            qpd[q]->density = function.value(fe_values.quadrature_point(q));
+        }
+  }
+
+  ConstraintMatrix constraints;
+  constraints.close ();
+
+  LA::MPI::Vector field(dof_handler.locally_owned_dofs(),
+                        MPI_COMM_WORLD);
+  VectorTools::project_distributed<dim,LA::MPI::Vector,dim,fe_degree>
+  (MappingQGeneric<dim>(1),
+   dof_handler,
+   constraints,
+   quadrature_formula,
+   [=] (const typename DoFHandler<dim>::active_cell_iterator & cell, const unsigned int q) -> double { return qp_manager.get_data(cell)[q]->density; },
```

Comment:
@bangerth :  here is a unit test which needs `std::function` to project quadrature point data on the FE space. This can be used in visualization, a-posteriori error estimates and other, usually not to computationally intensive, tasks.  Same below for quadrature data in matrix free, where a function returns `VectorizedArray<number>`.